chalice/awsclient.py

```diff
@@ -102,18 +133,45 @@ def _call_client_method_with_retries(self, method, kwargs):
         while True:
             try:
                 response = method(**kwargs)
-            except client.exceptions.InvalidParameterValueException:
+            except client.exceptions.InvalidParameterValueException as e:
                 # We're assuming that if we receive an
                 # InvalidParameterValueException, it's because
                 # the role we just created can't be used by
                 # Lambda so retry until it can be.
                 self._sleep(self.DELAY_TIME)
                 attempts += 1
-                if attempts >= self.LAMBDA_CREATE_ATTEMPTS:
+                message = e.response['Error'].get('Message', '')
+                # Do not retry if exceeded retries or does not have anything
+                # related to IAM roles.
+                if attempts >= self.LAMBDA_CREATE_ATTEMPTS or \
+                        not re.search('role.*cannot be assumed', message):
                     raise
                 continue
             return response
 
+    def _get_lambda_code_deployment_error(self, error, context):
+        # type: (Any, LambdaErrorContext) -> LambdaClientError
+        error_cls = LambdaClientError
+        if (isinstance(error, RequestsConnectionError) and
+                context.deployment_size > MAX_LAMBDA_DEPLOYMENT_SIZE):
+            # When the zip deployment package is too large and Lambda
+            # aborts the connection as chalice is still sending it
+            # data
+            error_cls = DeploymentPackageTooLargeError
+        elif isinstance(error, ClientError):
+            code = error.response['Error'].get('Code')
```

Comment:
These will default to `None` which will crash in your `__contains__` call on L169.  These should default to `str` types, probably just the empty string.